src/librustc_ast/ast.rs

```diff
@@ -309,19 +310,48 @@ pub type GenericBounds = Vec<GenericBound>;
 /// Specifies the enforced ordering for generic parameters. In the future,
 /// if we wanted to relax this order, we could override `PartialEq` and
 /// `PartialOrd`, to allow the kinds to be unordered.
-#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
+#[derive(PartialEq, Eq, Hash, Clone, Copy)]
 pub enum ParamKindOrd {
     Lifetime,
     Type,
-    Const,
+    // unordered is only true if sess.has_features().const_generics
+    // is active. Specifically, if it's only `min_const_generics`, it will still require ordered
+    // types.
+    Const { unordered: bool },
+}
+
+impl PartialOrd for ParamKindOrd {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        use ParamKindOrd::*;
+        let ord = match self {
```

Comment:
You can use `match (self, other)` here which is cleaner in my opinion.

Using `feature(or_patterns)` we can also write the following here:
```rust
(
    ParamKindOrd::Type | ParamKindOrd::Const { unordered: true },
    ParamKindOrd::Type | ParamKindOrd::Const { unordered: true }
) => true
```